add erb_rtk message

justprosh opened the pull request.

egorf commented:
@amilcarlucas Can you please take a look at this? We feel adding additional info like this will be a lot cleaner.
 
We also need clarification regarding the following fields:
* GPS Week Number of last baseline. Does **last baseline** mean the last base observation?
* Current number of integer ambiguity hypotheses. I am not sure what's intended to be here as well.

amilcarlucas commented:
Does last baseline mean the last base observation?
yes.

Current number of integer ambiguity hypotheses:
https://books.google.de/books?id=FQlhBgAAQBAJ&pg=PA336&lpg=PA336&dq=GPS+number+of+integer+ambiguity+hypotheses&source=bl&ots=-K0nJOu4AC&sig=gapE5S7MJbwKDz0sNY2iAHagnLk&hl=de&sa=X&ved=0ahUKEwit_Oe20OjVAhWCZFAKHUuZBTU4ChDoAQgtMAE#v=onepage&q=GPS%20number%20of%20integer%20ambiguity%20hypotheses&f=false

Usually the uBlox GPS reports this one directly

amilcarlucas commented on src/erb.c:
```diff
@@ -226,6 +243,46 @@ static int buildsvi(char *payload, struct erb_svi_head sviHead, struct erb_svi_s
+        for (i = 0; i < 3; i++) 
+            baseline_enu[i] = sol->rr[i];
+    }
+    rtk.base_num_sats = sol->ns;    /* Current number of satellites used for RTK calculation */
```
This should be the number of satelites that both rover and base share (the sats that are visible by both receivers at the same time) and that get used in calculating the RTK solution.

This should not be the number of satelites visible by the rover (status.numSV).

So either correct this one or correct status.numSV

egorf replied:
I think the `sol->ns` is RTK-related info, so it'll stay here. The STAT message will be fixed to send `numSV` as more general GNSS info and to avoid duplication. Good point.

amilcarlucas commented on src/erb.c:
```diff
@@ -226,6 +243,46 @@ static int buildsvi(char *payload, struct erb_svi_head sviHead, struct erb_svi_s
+    time_sec = (double)sol->time.time + sol->time.sec - sol->age;
+    time_of_last_baseline.time = (time_t)time_sec;
+    time_of_last_baseline.sec = time_sec - (double)time_of_last_baseline.time;
+    tow_ms = time2gpst(time_of_last_baseline, &week_number);
```
Did you test this ? It looks wrong.

justprosh replied:
Yes, it works correctly

amilcarlucas commented on src/erb.c:
```diff
@@ -226,6 +243,46 @@ static int buildsvi(char *payload, struct erb_svi_head sviHead, struct erb_svi_s
+
+    rtk.base_week_number = week_number;  /* GPS Week Number of last baseline */
+    rtk.base_time_week_ms = tow_ms;  /* GPS Time of Week of last baseline in milliseconds */
+    rtk.iar_num_hypotheses = 0;  /* TODO Current number of integer ambiguity hypotheses */
```
This one is available in uBlox firmware. RTKLIB should also calculate something similar to this one.

egorf replied:
We don't have to follow u-blox protocol exactly. What's the use for this anyway?

amilcarlucas commented on src/erb.c:
```diff
@@ -226,6 +243,46 @@ static int buildsvi(char *payload, struct erb_svi_head sviHead, struct erb_svi_s
+    rtk.base_week_number = week_number;  /* GPS Week Number of last baseline */
+    rtk.base_time_week_ms = tow_ms;  /* GPS Time of Week of last baseline in milliseconds */
+    rtk.iar_num_hypotheses = 0;  /* TODO Current number of integer ambiguity hypotheses */
+    rtk.num_leap_seconds = 0x7F; /* TODO access rtcm->nav->leaps instead; GPS leap seconds (0x7F indicates invalid) */
```
This one is still missing. I know that it only gets updated once every 12,5 minutes.... but it can get initialized with 7F and once data is available the value can be updated.

amilcarlucas commented on src/solution.c:
```diff
@@ -1632,7 +1632,7 @@ extern int outsols(unsigned char *buff, const sol_t *sol, const double *rb,
         case SOLF_NMEA: p+=outnmea_rmc(p,sol);
                         p+=outnmea_gga(p,sol);     break;
-        case SOLF_ERB:  p+=outerb(p,sol);          break;
+        case SOLF_ERB:  p+=outerb(p,sol,rb);          break;
```
Indentation. On my patch this was correctly indented :)

amilcarlucas reviewed (changes requested):
I still do not like the idea of a separated message, but if you fix the stuff I pointed out here, I'll do the required changes on the ardupilot side.

amilcarlucas commented:
Please compare my response time with your response time :)

amilcarlucas commented:
Merged your change to my PR and added credits to you:
https://github.com/emlid/RTKLIB/pull/21/commits/9075a2f9ee363fa8532a6e234eebb5abd949bd62

egorf commented:
> It took me less then 38 minutes to respond !

You are amazing. 

Understand that this PR is just a draft, so we can talk it over with you. Since generally you are ok with it, we will finish it.

amilcarlucas commented:
Line 217 is also TODO.

amilcarlucas commented:
Are you planing to add num_leap_seconds and iar_num_hypotheses on a 0.2.0 ERB protocol version, or are you planing to add another ERB message with those fields ?

egorf commented:
@amilcarlucas Do you mean this line?
```
    velocity.accS = 0x00;
```

Regarding the leap seconds and the hypothesis, you will need to justify why they should be included. Adding them just because same info is available in the u-blox protocol does not make much sense. Also, including these values will require extending the sol_t struct, which is kinda dirty.

amilcarlucas commented:
> velocity.accS = 0x00;

Yes I meant that line. The EKF needs velocity accuracy information.

Regarding the fields, the justification is the Mavlink protocol: http://mavlink.org/messages/common#GPS_RTK

Most of the fields are done now, but ... as you can see it also expects:

- rtk_receiver_id | uint8_t | Identification of connected RTK receiver.
- rtk_health | uint8_t | GPS-specific health report for RTK data.
- rtk_rate | uint8_t | Rate of baseline messages being received by GPS, in HZ
- accuracy | uint32_t | Current estimate of baseline accuracy.

I'm ignoring the first 3 ones for now, if you can add that, that would be great. :)
I'm abusing the protocol and I'm using RTKLIB's AR validation ratio as accuracy information in the fourth parameter. The other ardupilot team members do not like that. They would like to have a real accuracy estimate here, instead of an RTKLIB specific value with no physical accuracy (meters) relevance.

Now you guys come along and say that iar_num_hypotheses is uBlox/septentrio/topcon/trimble specific and do not want to implement it, so that is another field less

- iar_num_hypotheses | int32_t | Current number of integer ambiguity hypotheses.


num_leap_seconds is used by the ardupilot to convert from GPS time into UTC time. The value is currently hardcoded to 18 in the ardupilot code. But that is not a good solution, old FW will not give you correct time in 2019. And if someone forgets to update the code new FW version won't give correct time either.

I hope this is clear now. If you have any questions do not hesitate to ask.

amilcarlucas commented:
Nice that you rebased this ... but you do not give me credit in the comit message :(

egorf commented:
@amilcarlucas I totally understand your desire to have some value which would allow to determine the general health of RTK solution. The problem is, however, that all the numbers we've discussed, represent the internal state and are not a reliable way to check whether the RTK solution is "good". At the moment values like AR ratio are only suitable for human monitoring. We might, however, come up with something like this in the future.

justprosh committed:
src: erb.c: fix ERB_STAT and add ERB_RTK messages

This PR is a remake of PR #21 by @amilcarlucas.
Change sol->ns on sol->nSV and add modified version of ERB_RTK

egorf commented:
@amilcarlucas The commit message has been fixed and the code tested. I think we are ready to merge.

amilcarlucas commented:
Thank you for implementing this.
velocity.accS is still a TODO.

amilcarlucas commented:
I did a WIP PR for ardupilot https://github.com/ArduPilot/ardupilot/pull/6858
I will do some testing in the next few days

egorf commented:
@amilcarlucas Thank you
